runtime/lua/vim/treesitter/languagetree.lua

```diff
@@ -409,6 +428,46 @@ function LanguageTree:_add_injections()
   return query_time
 end
 
+--- Run an asynchronous parse, calling {on_parse} when complete.
+---
+--- @param range boolean|Range|nil
+--- @param on_parse fun(trees: TSTrees)
+function LanguageTree:_async_parse(range, on_parse)
+  coroutine.resume(
+    --- @param co_range boolean|Range|nil
+    --- @param co_cb fun(trees: TSTrees)
+    coroutine.create(function(co_range, co_cb)
+      local co = coroutine.running()
+      local ct = self:_buf().changedtick
+      local unfinished = true
+      --- @type TSTrees
+      local trees
+
+      while unfinished do
+        -- If buffer was changed in the middle of parsing, reset parse state
+        if self:_buf().changedtick ~= ct then
+          self._parser:reset()
+          self:invalidate()
+          coroutine.yield()
+        end
+
+        trees, unfinished = self:_parse(co_range, default_parse_timeout_ms)
+
+        if unfinished then
+          vim.schedule(function()
+            coroutine.resume(co)
+          end)
+          coroutine.yield()
+        else
+          co_cb(trees)
```

Comment:
This should be called outside of the loop as the final statement. Also we need to make sure this isn't called in the coroutine so should probably wrap with vim.schedule.

Reply:
I had wrapped in `vim.schedule` initially and noticed lots of highlight flickering due to the slow injection code which takes place (I can upload a screencast). Running without schedule removes all flicker. Is it not safe to just run immediately here (no schedule) since we assume we got here by a `vim.schedule` in the first place?

Reply:
Ah this is because you pass the callback in on_win. In that case if we ever go pass the segment time there will be some stutter, and the common case is we never reach the parse timeout, so there is no stutter.

In that case can we make it so the first call to parse runs outside of the coroutine and subsequent iterations happen within the coroutine, then we should be able to wrap the callback in schedule. If the first iteration of parse returns a tree then we can call the callback without schedule.

Hope that makes sense.

Reply:
Sorry, it's still not clear to me what problem this is trying to solve. It's an issue if the callback is run within the coroutine? I haven't found any problems so far with the current implementation

> EDIT: or there might be some way to write this so parse is always run on the main thread. Might even be possible to do without a coroutine.

Is this not the definition of just running a synchonous parse?

Reply:
Here's an implementation that doesn't use coroutines.

```lua
--- Run an asynchronous parse, calling {on_parse} when complete.
---
--- @param range boolean|Range|nil
--- @param on_parse fun(trees?: TSTrees, err?: string)
function LanguageTree:_async_parse(range, on_parse)
  local ct = self:_buf().changedtick
  local total_parse_time = 0
  local redrawtime = vim.o.redrawtime

  local function step()
    -- If buffer was changed in the middle of parsing, reset parse state
    if self:_buf().changedtick ~= ct then
      self._parser:reset()
      self:invalidate()
	  ct = self:_buf().changedtick
      total_parse_time = 0
    end

    local start = vim.uv.hrtime()
    local trees, unfinished = self:_parse(range, default_parse_timeout_ms)
    total_parse_time = total_parse_time + ((vim.uv.hrtime() - start) / 1e9)

    if not unfinished then
      on_parse(trees)
    elseif total_parse_time > redrawtime then
      on_parse(nil, 'TIMEOUT')
    else
      vim.schedule(step)
    end
  end

  step()
end
```

Note I've changes the signature of `on_parse` to match the UV style more.

Reply:
Awesome, super cool that you got this working without coroutines, I couldn't figure it out myself.

One thing to bring up: do you think it's fine to have the tcall here as a time measurement? This will account for e.g. the slow injection processing taking ~1.5 seconds for the big linux file, whereas the parse time would still be on average 3ms. Before I had each iteration only add the parse timeout so as to now skew the time from other processing noise. Do you think we should include this additional timing data?

Reply:
Yes I think we should. The goal here is to limit the amount of time parsing blocks the main thread for, so this must included everything. Just doing it for `TSParser:parse()` is just the first step.